### ISSUE TITLE
CMake Error at cmake/modules/CMakeLists.txt

### ISSUE THREAD

**Issue by shoober420** [llvm/llvm-project#53281]
I'm not sure which commit caused this, but when compiling against LLVM master branch, I get the following errors. Full logs attached.

Build Log:
```
CMake Error at cmake/modules/CMakeLists.txt:1 (include):
  include could not find requested file:

    ExtendPath


CMake Error at cmake/modules/CMakeLists.txt:3 (include):
  include could not find requested file:

    FindPrefixFromConfig


CMake Error at cmake/modules/CMakeLists.txt:113 (find_prefix_from_config):
  Unknown CMake command "find_prefix_from_config".
```

CMakeError Log:
```
Run Build Command(s):/usr/bin/ninja cmTC_ba7f6 && [1/2] Building CXX object CMakeFiles/cmTC_ba7f6.dir/src.cxx.o
FAILED: CMakeFiles/cmTC_ba7f6.dir/src.cxx.o 
/usr/bin/x86_64-pc-linux-gnu-g++ -m32 -DHAS_WERROR_GLOBAL_CTORS   -DNDEBUG -O2 -flarge-source-files -w -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections    -Werror=global-constructors -std=c++14 -o CMakeFiles/cmTC_ba7f6.dir/src.cxx.o -c /var/tmp/portage/sys-devel/llvm-14.0.0.9999/work/llvm-14.0.0.9999_build-abi_x86_32.x86/CMakeFiles/CMakeTmp/src.cxx
cc1plus: error: '-Werror=global-constructors': no option '-Wglobal-constructors'; did you mean '-felide-constructors'?
ninja: build stopped: subcommand failed.


Source file was:
int main() { return 0; }
```

CMakeOutput Log:
```
Performing C++ SOURCE FILE Test CXX_SUPPORTS_FDATA_SECTIONS succeeded with the following output:
Change Dir: /var/tmp/portage/sys-devel/llvm-14.0.0.9999/work/llvm-14.0.0.9999_build-abi_x86_32.x86/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_b5c88 && [1/2] Building CXX object CMakeFiles/cmTC_b5c88.dir/src.cxx.o
[2/2] Linking CXX executable cmTC_b5c88

Source file was:
int main() { return 0; }
```

build.log - [6WV53QV2L.txt](https://github.com/llvm/llvm-project/files/7893626/6WV53QV2L.txt)

CMakeError.log - [99FFXC737.txt](https://github.com/llvm/llvm-project/files/7893629/99FFXC737.txt)

CMakeOutput.log - [HMHA288ZK.txt](https://github.com/llvm/llvm-project/files/7893634/HMHA288ZK.txt)

**Comment by jrprice** [comment]
If you're using a sparse-checkout, you'll need to add the recently added `/cmake` directory to your `.gito/info/sparse-checkout` file and update to get these additional CMake config files.

**Comment by shoober420** [comment]
@jrprice 

I'm on Gentoo, using the llvm-14.0.0.9999 ebuild. I don't even know where the "sparse-checkout" file is on my system. I don't seem to have a ".git" directory in my home directory or llvm directory.

**Comment by Mystro256** [comment]
Yeah I get this issue too. The checkout is fine, the files in /cmake/modules exist, but there's some sort of cmake related bug that isn't too clear.

**Comment by karineek** [comment]
I also got this issue. I checked recent changes in cmake/Modules/ folder. This commit looks related to it: https://github.com/llvm/llvm-project/commit/44e3365775101fec3fd355eda339282258d74415.
I first tried to upgrade my Cmake from 3.16.3 (ubuntu-20 default) to 3.22 (newest), but this didn't solve the issue.

I used its parent commit (git checkout b99833dc) to build llvm with coverage, and it seems to work (after turning off LLVM_INCLUDE_BENCHMARKS flag). It builds version 14.0.0.

**Comment by uaex** [comment]
maybe lost

**Comment by uaex** [comment]
@tstellar , some files not found:
```
llvm/cmake/modules/ExtendPath.cmake
llvm/cmake/modules/FindPrefixFromConfig.cmake
```
or using external files outside the repo.

**Comment by aaronpuchert** [comment]
> I also got this issue. I checked recent changes in cmake/Modules/ folder. This commit looks related to it: 44e3365775101fec3fd355eda339282258d74415.

Seems likely. @Ericson2314, why put these files in cmake instead of llvm/cmake? Are there projects that don't have access to the latter?

@tstellar, if this stands I think we'll need to provide a tarball for this cmake top-level directory as well, otherwise none of the subproject tarballs are usable anymore.

**Comment by aaronpuchert** [comment]
> @Ericson2314, why put these files in cmake instead of llvm/cmake?

Actually the [readme](../blob/main/cmake/README.rst) answers that, but I'm not sure I understand:

> The advantage of not putting these modules in an existing location like llvm/cmake/modules is two-fold:
>
> * Since they are not installed, we don't have to worry about any out-of-tree downstream usage, and thus there is no need for stability.

Fair enough, but who is forcing us to install all files in `llvm/cmake/modules`? The installation already excludes `GetHostTriple.cmake`.

> * Since they are available as part of the source at build-time, we don't have to do the usual stand-alone vs combined-build dances, avoiding much complexity.

They are not part of e.g. the [llvm subproject tarball](https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.0-rc1/llvm-14.0.0rc1.src.tar.xz) though. That used to be self-contained prior to LLVM 14.

**Comment by tstellar** [comment]
@aaronpuchert This solution exists in the main branch: https://reviews.llvm.org/D118481  Is this OK or would you prefer to have a separate cmake tarball?

**Comment by aaronpuchert** [comment]
@tstellar, thanks for pointing that out. I guess it came to late for rc1? I've worked with a separate tarball for now, but this should also work, perhaps even without changes on my side. It would give me multiple copies of that directory, but that should be bearable since it's pretty small.

**Comment by h-vetinari** [comment]
> This solution exists in the main branch: https://reviews.llvm.org/D118481

That's great news, thanks! Does 32a0482a65b8 need to be marked for backport in a separate issue, or is this one enough?

**Comment by tstellar** [comment]
/cherry-pick https://github.com/llvm/llvm-project/commit/32a0482a65b86cf0d84ea7e784cca9852df5c67c

**Comment by Ericson2314** [comment]
Glad `/cmake` is going in the tarballs now.

**Comment by h-vetinari** [comment]
> Please manually backport the fix and push it to your github fork. Once this is done, please add a comment like this:

/branch h-vetinari/llvm-project/backport_D118481

**Comment by h-vetinari** [comment]
@tstellar
the backport cherry-pick is ready here (leaving a comment since it missed rc2 despite being milestoned, hoping it can get into rc3).

**Comment by tstellar** [comment]
Merged: 1e3d5ccab6b616baa91448462951b43f28bd5601

**Comment by h-vetinari** [comment]
This still seems to be an issue with rc4:
```
CMake Error at cmake/modules/CMakeLists.txt:1 (include):
  include could not find requested file:

    ExtendPath


CMake Error at cmake/modules/CMakeLists.txt:3 (include):
  include could not find requested file:

    FindPrefixFromConfig


CMake Error at cmake/modules/CMakeLists.txt:113 (find_prefix_from_config):
  Unknown CMake command "find_prefix_from_config".
```

**Comment by h-vetinari** [comment]
I'm hitting this with both `llvm` and `compiler-rt`, so the fix didn't seem to take effect properly... Can we reopen this issue, or should I raise a new one?

**Comment by tstellar** [comment]
@h-vetinari Can you provide instructions for reproducing this?

**Comment by mati865** [comment]
@tstellar one of the easiest examples is building compiler-rt on Linux or Windows (tested with mingw-w64):
```
$ cd compiler-rt
$ mkdir build && cd build/
$ cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ..
-- The C compiler identification is Clang 12.0.0
-- The CXX compiler identification is Clang 12.0.0
-- The ASM compiler identification is Clang
-- Found assembler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:31 (include):
  include could not find load file:

    SetPlatformToolchainTools


CMake Error at cmake/base-config-ix.cmake:9 (include):
  include could not find load file:

    ExtendPath
Call Stack (most recent call first):
  CMakeLists.txt:32 (include)


-- Looking for unwind.h
-- Looking for unwind.h - found
-- Looking for rpc/xdr.h
-- Looking for rpc/xdr.h - found
CMake Error at cmake/base-config-ix.cmake:108 (extend_path):
  Unknown CMake command "extend_path".
Call Stack (most recent call first):
  CMakeLists.txt:32 (include)


-- Configuring incomplete, errors occurred!
See also "/tmp/compiler-rt/build/CMakeFiles/CMakeOutput.log".
See also "/tmp/compiler-rt/build/CMakeFiles/CMakeError.log".
```

**Comment by h-vetinari** [comment]
Yup, something like @mati865's example is similar to what we're doing (i.e. quite vanilla).

**Comment by h-vetinari** [comment]
Looking some more at https://github.com/llvm/llvm-project/commit/1e3d5ccab6b616baa91448462951b43f28bd5601, I don't think the following works as intended:
```
        git archive --prefix=$proj-$release$rc.src/ -o $target_archive_file.tmp $tree_id .
        # Get relative path to top-level cmake directory to be packaged
        # alongside the project. Append that path to the tarball.
        cmake_rel_path=$(realpath --relative-to=. $llvm_src_dir/cmake)
        tar --append -f $target_archive_file.tmp $cmake_rel_path
```

In particular, I don't think that tar is able to `--append` a folder. Did you test this somehow @kwk?

AFAICT from the [documentation](https://git-scm.com/docs/git-archive) of `git archive`, it should be possible to add several paths, e.g. something like
```
        pushd $llvm_src_dir/$proj
        git archive --prefix=$proj-$release$rc.src/ $tree_id . $llvm_src_dir/cmake | xz >$target_dir/$(template_file $proj)
```

**Comment by h-vetinari** [comment]
Could someone reopen this issue please?

I'd prefer to keep the context of the discussion so far, but if not, I'll just have to open a new one...

**Comment by aaronpuchert** [comment]
> Looking some more at [1e3d5cc](https://github.com/llvm/llvm-project/commit/1e3d5ccab6b616baa91448462951b43f28bd5601), I don't think the following works as intended:
> 
> ```
>         git archive --prefix=$proj-$release$rc.src/ -o $target_archive_file.tmp $tree_id .
>         # Get relative path to top-level cmake directory to be packaged
>         # alongside the project. Append that path to the tarball.
>         cmake_rel_path=$(realpath --relative-to=. $llvm_src_dir/cmake)
>         tar --append -f $target_archive_file.tmp $cmake_rel_path
> ```

Another issue is that it uses the checked-out source tree instead of taking files from tag/commit passed as argument.

> AFAICT from the [documentation](https://git-scm.com/docs/git-archive) of `git archive`, it should be possible to add several paths, e.g. something like
> 
> ```
>         pushd $llvm_src_dir/$proj
>         git archive --prefix=$proj-$release$rc.src/ $tree_id . $llvm_src_dir/cmake | xz >$target_dir/$(template_file $proj)
> ```

But would it appear in the right place then?

How about using `git archive` to package up the `cmake/` directory with the correct prefix, and then use `tar --concatenate` to append it to the other archives?

**Comment by aaronpuchert** [comment]
Posted [D121972](https://reviews.llvm.org/D121972), feel free to test it.

**Comment by kwk** [comment]
> Looking some more at [1e3d5cc](https://github.com/llvm/llvm-project/commit/1e3d5ccab6b616baa91448462951b43f28bd5601), I don't think the following works as intended:
> 
> ```
>         git archive --prefix=$proj-$release$rc.src/ -o $target_archive_file.tmp $tree_id .
>         # Get relative path to top-level cmake directory to be packaged
>         # alongside the project. Append that path to the tarball.
>         cmake_rel_path=$(realpath --relative-to=. $llvm_src_dir/cmake)
>         tar --append -f $target_archive_file.tmp $cmake_rel_path
> ```
> 
> In particular, I don't think that tar is able to `--append` a folder. Did you test this somehow @kwk?

I tested it and it currently runs every day. Take any tar ball in this release to see that there's a `cmake` top-level directory.

> 
> AFAICT from the [documentation](https://git-scm.com/docs/git-archive) of `git archive`, it should be possible to add several paths, e.g. something like
> 
> ```
>         pushd $llvm_src_dir/$proj
>         git archive --prefix=$proj-$release$rc.src/ $tree_id . $llvm_src_dir/cmake | xz >$target_dir/$(template_file $proj)
> ```

I don't know why I didn't see this or if I didn't test this. It sounds like a good idea.

**Comment by kwk** [comment]
> Another issue is that it uses the checked-out source tree instead of taking files from tag/commit passed as argument.

That indeed is an issue.

**Comment by aaronpuchert** [comment]
@tstellar, need a backport for this?

**Comment by h-vetinari** [comment]
Thanks everyone! I'd love to test this, but it's not so easy (I'd have to create the tarballs in CI per platform and I'd have to work around a couple thing). The best thing would be another RC obviously, but I'm guessing that's not in the cards anymore. In any case, I'm (unsurprisingly) in favour of this being backported.

**Comment by tstellar** [comment]
Merged: add3ab7f4c8a 3f43d803382d 23d08271a4b2

**Comment by uaex** [comment]
```
CMake Error at cmake/modules/CMakeLists.txt:1 (include):
  include could not find requested file:

    ExtendPath


CMake Error at cmake/modules/CMakeLists.txt:3 (include):
  include could not find requested file:

    FindPrefixFromConfig


CMake Error at cmake/modules/CMakeLists.txt:113 (find_prefix_from_config):
  Unknown CMake command "find_prefix_from_config".


Configuring incomplete, errors occurred!
```

I'm puzzled about this issue. It seems that `ExtendPath.cmake` and `FindPrefixFromConfig.cmake` does not exist in repository, and the solution does not resolve it. 😞

**Comment by aaronpuchert** [comment]
> It seems that `ExtendPath.cmake` and `FindPrefixFromConfig.cmake` does not exist in repository, and the solution does not resolve it.

They definitely exist in the repository: in [cmake/Modules/ExtendPath.cmake](https://github.com/llvm/llvm-project/blob/llvmorg-14.0.0/cmake/Modules/ExtendPath.cmake) and [cmake/Modules/FindPrefixFromConfig.cmake](https://github.com/llvm/llvm-project/blob/llvmorg-14.0.0/cmake/Modules/FindPrefixFromConfig.cmake). They weren't packaged into the subproject tarballs with the release candidates, but this is fixed with 14.0.0.

**Comment by SupervisedThinking** [comment]
still missing / wrong packed in https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.0/llvm-14.0.0.src.tar.xz

**Comment by heitbaum** [comment]
> still missing / wrong packed in https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.0/llvm-14.0.0.src.tar.xz

I included the missing in the LE PR. Assuming that 14.0.1 should be right - https://github.com/heitbaum/LibreELEC.tv/commit/138e929c4a7c0d4d9d1d56a9e9a1f24504fa4ad2

**Comment by aaronpuchert** [comment]
> still missing / wrong packed in https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.0/llvm-14.0.0.src.tar.xz

Can you please run and compare:
```
> tar --list -f llvm-14.0.0.src.tar.xz | tail
cmake/
cmake/Modules/
cmake/Modules/EnableLanguageNolink.cmake
cmake/Modules/ExtendPath.cmake
cmake/Modules/FindPrefixFromConfig.cmake
cmake/Modules/HandleCompilerRT.cmake
cmake/Modules/HandleOutOfTreeLLVM.cmake
cmake/Modules/LLVMCheckCompilerLinkerFlag.cmake
cmake/Modules/SetPlatformToolchainTools.cmake
cmake/README.rst
```
For me both `ExtendPath` and `FindPrefixFromConfig` are there.

**Comment by SupervisedThinking** [comment]
as I said they are wrong packed: 

![Bildschirmfoto vom 2022-04-10 17-25-12](https://user-images.githubusercontent.com/36541229/162626882-f1fa3efb-0071-46c7-a93c-de9afd84279d.png)

They were added to the root directory and not where they belong:

![Bildschirmfoto vom 2022-04-10 17-27-40](https://user-images.githubusercontent.com/36541229/162626958-4d856016-9fad-4372-ad6e-7f601b781c5e.png)

**Comment by aaronpuchert** [comment]
These are two different folders in the repository: `cmake/modules` and `llvm/cmake/modules`. The tarball simply reflects that. Now I'm not sure we need that separate directory, but it was decided to have it to separate files that should be installed from files that shouldn't. (See my [earlier comment](https://github.com/llvm/llvm-project/issues/53281#issuecomment-1052686976).)

As for `modules` vs. `Modules`, CMake doesn't care. Not sure why we went with different capitalization, but that ship sailed a long time ago:
```
cmake/Modules
clang/cmake/modules
compiler-rt/cmake/Modules
flang/cmake/modules
libc/cmake/modules
libcxxabi/cmake/Modules
libcxx/cmake/Modules
libunwind/cmake/Modules
lldb/cmake/modules
lld/cmake/modules
llvm/cmake/modules
mlir/cmake/modules
```
This bug here is simply a packaging issue. If you think we shouldn't have that top-level `cmake` directory, better open a separate bug. But read that [README](https://github.com/llvm/llvm-project/blob/main/cmake/README.rst) first.

**Comment by sbourdeauducq** [comment]
Bug is still present in release 14.0.3.

**Comment by aaronpuchert** [comment]
Could you be a bit more specific what your setup is? The subproject tarballs contain the necessary files now.

**Comment by SupervisedThinking** [comment]
@aaronpuchert I guess a 14.0.4 will be fixed then? Or did you re-release the 14.0.3 files with https://github.com/PBHDK/llvm-project/commit/4959d294267ad5b073582f1b00a6c77665b8b863

**Comment by aaronpuchert** [comment]
The bug is closed because by our understanding the issue has been fixed with 14.0.0, and only release candidates were affected. So what is the issue that you're seeing? The files are there, so somehow the build that you're doing doesn't pick them up.

**Comment by sbourdeauducq** [comment]
Well, just open the 14.0.3 release tarball (https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.3/llvm-14.0.3.src.tar.xz). There is a ``cmake`` folder at the top level which shouldn't be there in addition to the legitimate ``llvm-14.0.3.src`` folder, with consequences as described in this Issue. Other tarballs are also affected.

**Comment by heitbaum** [comment]
Using `llvm-project` (14) source now versus `llvm` (13), this addresses the change  of packaging.
- https://github.com/LibreELEC/LibreELEC.tv/pull/6492

**Comment by tstellar** [comment]
Here is a proposed change to remove the cmake bundling from each tarball and package it separately: https://reviews.llvm.org/D125798
To me this seems like the best option for now.  It will get us back into our original state with the tarballs, and then after that we can discuss other improvements like the directory names, etc.

**Comment by heitbaum** [comment]
> Here is a proposed change to remove the cmake bundling from each tarball and package it separately: https://reviews.llvm.org/D125798
> To me this seems like the best option for now. It will get us back into our original state with the tarballs, and then after that we can discuss other improvements like the directory names, etc.

As a packaging manager - I would ask that each `.tar.gz` file can discretely be built in its directory, without the `..` syntax. 

Whilst the updated `package.mk` file in LibreELEC uses the override of `llvm/CMakeLists.txt` - I would have preferred an `CMakeLists.txt` in the root directory which would have called `llvm/CMakeLists.txt`.

I would support NOT having 2 parent directories in the tar.gz file as `llvm-14` is being shipped. If this is the need - then suggest moving both llvm (remove the version) and cmake as subdirectories in the tar.gz:llvm-14.0.3/ directory.

**Comment by aaronpuchert** [comment]
> There is a `cmake` folder at the top level which shouldn't be there in addition to the legitimate `llvm-14.0.3.src` folder, with consequences as described in this Issue.

The issue described here stems from the `cmake` directory **not** being there, because the files simply didn't exist. With the directory they should be in `../cmake/Modules` as in the git tree and thus should be found.

> I would support NOT having 2 parent directories in the tar.gz file as `llvm-14` is being shipped. If this is the need - then suggest moving both llvm (remove the version) and cmake as subdirectories in the tar.gz:llvm-14.0.3/ directory.

Isn't that #54716 instead of this issue?

**Comment by sshedi** [comment]
This issue present in llvm-15.0.1 as well.

**Comment by sshedi** [comment]
Also, this is throwing error:
https://github.com/llvm/llvm-project/blob/329fda39c507e8740978d10458451dcdb21563be/llvm/CMakeLists.txt#L329

With file not found which is right. There is no `third-party/benchmark` directory in the mentioned location in the released tarball.
I'm aware that this is a separate issue and should be discussed in a different thread but mentioning it here anyway because this is also a build regression.

**Comment by aaronpuchert** [comment]
> This issue present in llvm-15.0.1 as well.

With LLVM 15 you'll have to download the `cmake` project tarball and unpack it alongside whatever project you're building.

> Also, this is throwing error:
> 
> https://github.com/llvm/llvm-project/blob/329fda39c507e8740978d10458451dcdb21563be/llvm/CMakeLists.txt#L329
> 
> With file not found which is right. There is no `third-party/benchmark` directory in the mentioned location in the released tarball. I'm aware that this is a separate issue and should be discussed in a different thread but mentioning it here anyway because this is also a build regression.

I'm quite sympathetic, but large parts of the community don't care anymore about being able to build projects separately without checking out the whole monorepo.

> Edit: And workaround to bypass this error is using `-DLLVM_INCLUDE_BENCHMARKS=OFF` during cmake invocation.

Exactly. Usually there is no strong reason for these interdependencies and there is a way to pull them apart.

**Comment by sshedi** [comment]
FWIW, we consume source tarballs from released assets and build rpm from it. We don't want the full repo for that purpose and we start with the assumption that released assets are good enough to build rpms without much hassle.

**Comment by NobinPegasus** [comment]
I'm very new to LLVM, cmake and all that for some project I need to modify the LLVM and build it.
But while trying to build it I and facing this exact error. Can someone explain me in simpler words why is it happening?
And how to fix it. I'm using the latest version of LLVM.

**Comment by sshedi** [comment]
> I'm very new to LLVM, cmake and all that for some project I need to modify the LLVM and build it. But while trying to build it I and facing this exact error. Can someone explain me in simpler words why is it happening? And how to fix it. I'm using the latest version of LLVM.

Pass this while invoking cmake -DLLVM_INCLUDE_BENCHMARKS=OFF

**Comment by NobinPegasus** [comment]
I had actually used this pass -DLLVM_INCLUDE_BENCHMARKS=OFF but it still shows the error

**Comment by aaronpuchert** [comment]
Which error? Note that `-DLLVM_INCLUDE_BENCHMARKS=OFF` doesn't solve the error this issue is about, but an [unrelated error](https://github.com/llvm/llvm-project/issues/53281#issuecomment-1259755638).

If you see this error with the project tarballs, note that the packaging changed with LLVM 15. We're now packaging the top-level cmake directory separately, so you'll need to download that as well and unpack it next to the `llvm` tarball.

**Comment by NobinPegasus** [comment]
The following error
I'm using the CCR [build.sh](https://github.com/kevinkoo001/CCR/blob/master/build.sh) for building the LLVM

```
CMake Error at cmake/modules/CMakeLists.txt:1 (include):
  include could not find requested file:

    ExtendPath


CMake Error at cmake/modules/CMakeLists.txt:3 (include):
  include could not find requested file:

    FindPrefixFromConfig


CMake Error at cmake/modules/CMakeLists.txt:116 (find_prefix_from_config):
  Unknown CMake command "find_prefix_from_config".
```

**Comment by aaronpuchert** [comment]
> I'm using the CCR [build.sh](https://github.com/kevinkoo001/CCR/blob/master/build.sh) for building the LLVM

Can't find the setup of the sources there, so I assume you have done that manually. Where you unpack the `llvm` tarball you need to unpack the `cmake` tarball next to it and rename the top-level directory into `cmake`. You can find both under e.g. https://github.com/llvm/llvm-project/releases/tag/llvmorg-15.0.7.

**Comment by NobinPegasus** [comment]
should I rename  ` cmake-15.0.7.src.tar.xz ` to cmake? after extracting? 
What you mean it extract `llvm` and `cmake` to the same directory and then rename their parent directory as `cmake`

**Comment by aaronpuchert** [comment]
> Should I rename `cmake-15.0.7.src.tar.xz` to `cmake`? after extracting?

Yes.

> What you mean it extract `llvm` and `cmake` to the same directory and then rename their parent directory as `cmake`

After extraction into the same directory you should get two directories: `llvm-15.0.7.src` and `cmake-15.0.7.src`. The latter needs to be renamed `cmake`. The name of the former shouldn't matter, as does the name of the parent directory.

**Comment by beta-ziliani** [comment]
I got this issue trying to compile llvm-15.0.7 and the [solution proposed](https://github.com/llvm/llvm-project/issues/53281#issuecomment-1399635906) in the previous comment doesn't apply (there are no separate directories, and `./cmake` is part of the extracted `llvm-15.0.7.src` dir).

I also posted the issue in [the forum](https://discourse.llvm.org/t/issues-compiling-15-0-7/69255).

**Comment by Gamer069** [comment]
> Bug is still present in release 14.0.3.

 bug is still present in release 16.0.0

**Comment by cloudren2017** [comment]
From the master branch，copy an ExtendPath.cmake and a FindPrefixFromConfig.cmake from the dir cmake/Modules/ to your relative path cmake/modules/.
Remake and solve this problem.
